README.md
# World Control
Adds a shader setup which allows more control when using HDR/EXR lighting. Based on Lech Sokolowski (Chocofur) video BCON19.

!['Example NodeGroup'](https://raw.githubusercontent.com/wiki/schroef/World_Control/images/world_control_v0062.png?20211113)


## Workflow

Take a look at this video and see how this menu works

[!['Workflow using World Control'](https://raw.githubusercontent.com/wiki/schroef/World_Control/images/howto_lech_youtube.png?v28102021)](https://youtu.be/FF21foXMv4E?t=838)


### System Requirements

| **OS** | **Blender** |
| ------------- | ------------- |
| OSX | Blender 2.80+ |
| Windows | Blender 2.80+ |
| Linux | Not Tested |

### Installation Process

1. Download the latest <b>[release](https://github.com/schroef/World_Control/releases/)</b>
2. If you downloaded the zip file.
3. Open Blender.
4. Go to File -> User Preferences -> Addons.
5. At the bottom of the window, choose *Install From File*.
6. Select the file `World_Control_VERSION.zip` from your download location..
7. Activate the checkbox for the plugin that you will now find in the list.

### Changelog
[Full Changelog](CHANGELOG.md)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.0.6.2] - 2021-11-13
### Fixed
- Error in panel

## [0.0.6.1] - 2021-11-01
### Fixed
- Environment node uses "Environment" a very often, added this to check for old enviremont textures
- Correct default Saturation values in Node Group

## [0.0.6] - 2021-10-31
### Added
- auto-transfer old environment texture images > prevents from relinking them when switching modes

## [0.0.5] - 2021-10-30
### Changed
- World Control is now created in its own world named "WorldControl"

### Added
- Clean old setup when switching between advanced and basic
- WorldControl has its own world, so we dont mess with user own setup worlds. Easier for cleaning
- Easy switching operator, switch easy between basic and advanced

## [0.0.4] - 2021-10-30
### Changed
- Moved panel to View catergory
- Dynamic naming moved to panel itself > Panel Header has font size issue

## [0.0.3] - 2021-10-29
### Added
- Reset operator > set all inputs to default
- Panel view easy access from 3d viewport

## [0.0.2] - 2021-10-28
### Fixed 
- The Add function was visible in materials nodes as well, it added empty node in material, added poll to customnodecategrory 

### Changed
- Simplified node names on catergory, stripped "control"
- NodeGroup with was to narrow, added width

## [0.0.1] - 2021-10-27
### Added 
- Initial release repo 

## Notes
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
<!--### Official Rigify Info-->

[0.0.6.2]:https://github.com/schroef/World_Control/releases/tag/v0.0.6.2
[0.0.6.1]:https://github.com/schroef/World_Control/releases/tag/v0.0.6.1
[0.0.6]:https://github.com/schroef/World_Control/releases/tag/v0.0.6
[0.0.5]:https://github.com/schroef/World_Control/releases/tag/v0.0.5
[0.0.4]:https://github.com/schroef/World_Control/releases/tag/v0.0.4
[0.0.3]:https://github.com/schroef/World_Control/releases/tag/v0.0.3
[0.0.2]:https://github.com/schroef/World_Control/releases/tag/v0.0.2
[0.0.1]:https://github.com/schroef/World_Control/releases/tag/v0.0.1

wc_3dv_panel.py

import bpy, os, re
from bpy.props import StringProperty, EnumProperty, BoolProperty, IntProperty
from bpy.types import Panel, Operator, Menu


def engine_compatibility(context):
    scn = context.scene
    return scn.render.engine == 'CYCLES'

def find_basic(nt):
    return nt.nodes.find("WorldControlBasic")

def find_advanced(nt):
    return nt.nodes.find("WorldControlAdvanced")
    
def check_world_wc_check_mode(self, context):
    # Dirty method to get around 'RestrictContext'
    sceneLoaded = False
    try:
        if bpy.context.scene:
            sceneLoaded = True
    except:
        pass

    if sceneLoaded:
        scn = context.scene
        world = scn.world
        nt = world.node_tree
        if find_basic(nt) != -1:
            scn.wc_check_mode = "Basic"
        if find_advanced(nt) != -1:
            scn.wc_check_mode = "Advanced"
        else:
            scn.wc_check_mode = "World Control"

def check_world_control_type(context):
    world = bpy.context.scene.world
    nt = world.node_tree
    if find_basic(nt) != -1:
        return "World Control Basic"
    if find_advanced(nt) != -1:
        return "World Control Advanced"
    else:
        return "World Control"


def return_world_control_type():
    world = bpy.context.scene.world
    nt = world.node_tree
#    print(find_basic(nt) != -1)
    if find_basic(nt) != -1:
        return "WorldControlBasic"
    if find_advanced(nt) != -1:
        return "WorldControlAdvanced"
    else:
        return "World Control"


class WorldControlPanel:
    bl_space_type = "VIEW_3D"
    bl_region_type = "UI"
    bl_category = 'View'
    bl_options = {'DEFAULT_CLOSED'}


class WC_PT_main_panel(WorldControlPanel, Panel):          
    bl_label = "World Control"
    #bl_context = "objectmode"

    # wc_check_mode : StringProperty(name="ControlType",default = "Basic")

#     @classmethod
#     def poll(cls, context):
#         world = bpy.context.scene.world
#         nt = world.node_tree
#         basic = find_basic(nt)
#         advanced = find_advanced(nt)
# #        print(engine_compatibility(context))
# #        print(basic or advanced != -1)
#         return (basic != -1 or advanced != -1) and engine_compatibility(context)
    
    # Header text looks different vs regular BL_name > check with devs
    # def draw_header(self, context):
    #     layout = self.layout
    #     layout.label(text=str(check_world_control_type(context)))

        
    def draw(self, context):
        layout = self.layout
        layout.use_property_split = True
        layout.use_property_decorate = False
        
        scn = context.scene
        world = scn.world
        nodes = scn.world.node_tree.nodes
        nt = world.node_tree
        basic = find_basic(nt)
        advanced = find_advanced(nt)

        if (bpy.app.version[1] < 92):
            col = layout.column(align=True)
            col.label(text="Mode")
        else:
            col = layout.column(align=False, heading="Mode")
        col.use_property_decorate = False
        row = col.row(align=True)
        sub = row.row(align=True)

        if engine_compatibility and (basic != -1 or advanced != -1):
            sub.prop(scn,"wc_switch_control",expand = True)
            # sub.prop(scn, "wc_check_mode", emboss=False,icon='WORLD')
            # sub.label(text=str(check_world_wc_check_mode(context))) #icon='WORLD_DATA'
            reset_bg = row.operator("wc.reset_settings", icon = 'LOOP_BACK', text='')
            reset_bg.controlType = return_world_control_type()

            col = layout.column()
            col.separator()
            col.prop(nodes['Mapping'].inputs[2], "default_value", text = 'Rotation')   
        
        col = layout.column()
        if (basic == -1 and advanced == -1):   
            col.label(text = 'No World Control in shader', icon = 'INFO') 
        if not engine_compatibility:
            col.label(text = 'Not compatible with this render engine', icon = 'INFO')
            col.prop(scn.render, 'engine')
        # else:


class WC_PT_light_settings(WorldControlPanel, Panel):
    bl_category = "View"
    # bl_context = ".objectmode"  # dot on purpose (access from topbar)
    bl_label = "Light"
    bl_parent_id = "WC_PT_main_panel"
    bl_options = {'DEFAULT_CLOSED'}
    
    @classmethod
    def poll(cls, context):
        world = bpy.context.scene.world
        nt = world.node_tree
        basic = find_basic(nt)
        advanced = find_advanced(nt)
#        print(basic or advanced != -1)
        return (basic != -1 or advanced != -1) and engine_compatibility(context)

    def draw(self, context):
        layout = self.layout

        layout.use_property_split = True
        layout.use_property_decorate = False
        scn = context.scene
                 
        col = layout.column(align=True)                
        nodes = scn.world.node_tree.nodes
        controlType = return_world_control_type()  
        
#        col.prop(nodes['Mapping'].inputs[2], "default_value", text = 'Rotation')   
             
        if controlType == 'WorldControlBasic':
            col.prop(nodes['WorldControlBasic'].inputs[1], "default_value", text = 'Strength')
            col.prop(nodes['WorldControlBasic'].inputs[2], "default_value", text = 'Temperature')
            
            
        if controlType == 'WorldControlAdvanced':
            col.prop(nodes['WorldControlAdvanced'].inputs[1], "default_value", text = 'Strength')
            col.prop(nodes['WorldControlAdvanced'].inputs[2], "default_value", text = 'Saturation')
            col.prop(nodes['WorldControlAdvanced'].inputs[3], "default_value", text = 'Temperature')


class WC_PT_background_settings(WorldControlPanel, Panel):
    bl_category = "View"
    # bl_context = ".objectmode"  # dot on purpose (access from topbar)
    bl_label = "Background"
    bl_parent_id = "WC_PT_main_panel"
    bl_options = {'DEFAULT_CLOSED'}
    
    @classmethod
    def poll(cls, context):
        world = bpy.context.scene.world
        nt = world.node_tree
        basic = find_basic(nt)
        advanced = find_advanced(nt)
        return (basic != -1 or advanced != -1) and engine_compatibility(context)

    def draw(self, context):
        layout = self.layout

        layout.use_property_split = True
        layout.use_property_decorate = False
        scn = context.scene

        col = layout.column(align=True)                
        nodes = scn.world.node_tree.nodes
        
        controlType = return_world_control_type()  
     
        if controlType == 'WorldControlBasic':
            col.prop(nodes['WorldControlBasic'].inputs[3], "default_value", text = 'Strength')
            col.prop(nodes['WorldControlBasic'].inputs[4], "default_value", text = 'Saturation')
            
        if controlType == 'WorldControlAdvanced':
            col.prop(nodes['WorldControlAdvanced'].inputs[7], "default_value", text = 'Strength')
            col.prop(nodes['WorldControlAdvanced'].inputs[8], "default_value", text = 'Saturation') 


class WC_PT_diffuse_settings(WorldControlPanel, Panel):
    bl_category = "View"
    # bl_context = ".objectmode"  # dot on purpose (access from topbar)
    bl_label = "Diffuse"
    bl_parent_id = "WC_PT_main_panel"
    bl_options = {'DEFAULT_CLOSED'}
    
    @classmethod
    def poll(cls, context):
        world = bpy.context.scene.world
        nt = world.node_tree
        basic = find_basic(nt)
        advanced = find_advanced(nt)
        return (basic != -1 or advanced != -1) and engine_compatibility(context)

    def draw(self, context):
        layout = self.layout

        layout.use_property_split = True
        layout.use_property_decorate = False
        scn = context.scene

        col = layout.column(align=True)                
        nodes = scn.world.node_tree.nodes
        
        controlType = return_world_control_type()  
     
        if controlType == 'WorldControlBasic':
            col.prop(nodes['WorldControlBasic'].inputs[5], "default_value", text = 'Strength')
            col.prop(nodes['WorldControlBasic'].inputs[6], "default_value", text = 'Saturation')      
            
        if controlType == 'WorldControlAdvanced':
            col.prop(nodes['WorldControlAdvanced'].inputs[4], "default_value", text = 'Strength')
            col.prop(nodes['WorldControlAdvanced'].inputs[5], "default_value", text = 'Saturation')                         


class WC_PT_reflections_settings(WorldControlPanel, Panel):
    bl_category = "View"
    # bl_context = ".objectmode"  # dot on purpose (access from topbar)
    bl_label = "Reflections"
    bl_parent_id = "WC_PT_main_panel"
    bl_options = {'DEFAULT_CLOSED'}
    
    @classmethod
    def poll(cls, context):
        world = bpy.context.scene.world
        nt = world.node_tree
        basic = find_basic(nt)
        advanced = find_advanced(nt)
        return (advanced != -1) and engine_compatibility(context)

    def draw(self, context):
        layout = self.layout

        layout.use_property_split = True
        layout.use_property_decorate = False
        scn = context.scene

        col = layout.column(align=True)                
        nodes = scn.world.node_tree.nodes
        
        controlType = return_world_control_type()  
     
        if controlType == 'WorldControlAdvanced':
            col.prop(nodes['WorldControlAdvanced'].inputs[9], "default_value", text = 'Strength')
            col.prop(nodes['WorldControlAdvanced'].inputs[10], "default_value", text = 'Saturation') 


class WC_PT_glossy_settings(WorldControlPanel, Panel):
    bl_category = "View"
    # bl_context = ".objectmode"  # dot on purpose (access from topbar)
    bl_label = "Glossy"
    bl_parent_id = "WC_PT_main_panel"
    bl_options = {'DEFAULT_CLOSED'}
    
    @classmethod
    def poll(cls, context):
        world = bpy.context.scene.world
        nt = world.node_tree
        basic = find_basic(nt)
        advanced = find_advanced(nt)
        return (basic != -1) and engine_compatibility(context)

    def draw(self, context):
        layout = self.layout

        layout.use_property_split = True
        layout.use_property_decorate = False
        scn = context.scene

        col = layout.column(align=True)                
        nodes = scn.world.node_tree.nodes
        
        controlType = return_world_control_type()  
     
        if controlType == 'WorldControlBasic':
            col.prop(nodes['WorldControlBasic'].inputs[7], "default_value", text = 'Strength')
            col.prop(nodes['WorldControlBasic'].inputs[8], "default_value", text = 'Saturation')   


class WC_PT_refraction_settings(WorldControlPanel, Panel):
    bl_category = "View"
    # bl_context = ".objectmode"  # dot on purpose (access from topbar)
    bl_label = "Refraction"
    bl_parent_id = "WC_PT_main_panel"
    bl_options = {'DEFAULT_CLOSED'}
    
    @classmethod
    def poll(cls, context):
        world = bpy.context.scene.world
        nt = world.node_tree
        basic = find_basic(nt)
        advanced = find_advanced(nt)
        return (advanced != -1) and engine_compatibility(context)

    def draw(self, context):
        layout = self.layout

        layout.use_property_split = True
        layout.use_property_decorate = False
        scn = context.scene

        col = layout.column(align=True)                
        nodes = scn.world.node_tree.nodes
        
        controlType = return_world_control_type()  
        
        if controlType == 'WorldControlAdvanced':
            col.prop(nodes['WorldControlAdvanced'].inputs[11], "default_value", text = 'Strength')
            col.prop(nodes['WorldControlAdvanced'].inputs[12], "default_value", text = 'Saturation')            



def reset_settings(controlType):
    scn = bpy.context.scene
    worlds = bpy.data.worlds
    world = scn.world
    nodes = world.node_tree.nodes
    
    # Light
    if controlType == 'WorldControlBasic':
        nodes['WorldControlBasic'].inputs[1].default_value = 1.0 # Strength
        nodes['WorldControlBasic'].inputs[2].default_value = 5500 # Temperature

    if controlType == 'WorldControlAdvanced':
        nodes['WorldControlAdvanced'].inputs[1].default_value = 1.0 # Strength
        nodes['WorldControlAdvanced'].inputs[2].default_value = 1.0 # Saturation
        nodes['WorldControlAdvanced'].inputs[3].default_value = 5500 # Temperature

    # Background
    if controlType == 'WorldControlBasic':
        nodes['WorldControlBasic'].inputs[3].default_value = 1.0 # Strength
        nodes['WorldControlBasic'].inputs[4].default_value = 1.0 # Saturation
        
    if controlType == 'WorldControlAdvanced':
        nodes['WorldControlAdvanced'].inputs[7].default_value = 1.0 # Strength
        nodes['WorldControlAdvanced'].inputs[8].default_value = 1.0 # Saturation

    # Diffuse
    if controlType == 'WorldControlBasic':
        nodes['WorldControlBasic'].inputs[5].default_value = 1.0 # Strength
        nodes['WorldControlBasic'].inputs[6].default_value = 1.0 # Saturation
        
    if controlType == 'WorldControlAdvanced':
        nodes['WorldControlAdvanced'].inputs[4].default_value = 1.0 # Strength
        nodes['WorldControlAdvanced'].inputs[5].default_value = 1.0 # Saturation

    # Reflections
    if controlType == 'WorldControlAdvanced':
        nodes['WorldControlAdvanced'].inputs[9].default_value = 1.0 # Strength
        nodes['WorldControlAdvanced'].inputs[10].default_value = 1.0 # Saturation
    # Glossy
    if controlType == 'WorldControlBasic':
        nodes['WorldControlBasic'].inputs[7].default_value = 1.0 # Strength
        nodes['WorldControlBasic'].inputs[8].default_value = 1.0 # Saturation

    # Refraction
    if controlType == 'WorldControlAdvanced':
        nodes['WorldControlAdvanced'].inputs[11].default_value = 1.0 # Strength
        nodes['WorldControlAdvanced'].inputs[12].default_value = 1.0 # Saturation


class WC_OT_reset_settings(Operator):
    bl_idname = "wc.reset_settings"
    bl_label = "Reset"
    bl_description = "Reset to Default Values"
    bl_options = {'REGISTER', 'UNDO'}
    
    controlType : StringProperty(default = 'WorldControlBasic')
    
    # @classmethod
    # def poll(cls, context):
    #     return check_world_nodes() == 'OK'
    
    def execute(self, context):
        controlType = return_world_control_type()
        reset_settings(controlType)
        return {'FINISHED'}

classes = (
    WC_PT_main_panel,
    WC_PT_light_settings,
    WC_PT_background_settings,
    WC_PT_diffuse_settings,
    WC_PT_reflections_settings,
    WC_PT_glossy_settings,
    WC_PT_refraction_settings,
    WC_OT_reset_settings
    )
    

def register():
    from bpy.utils import register_class
    for cls in classes:
        register_class(cls)       

def unregister():
    from bpy.utils import unregister_class
    for cls in reversed(classes):
        unregister_class(cls)        



if __name__ == "__main__":
    register()        

__init__.py
'''
#############################################
# 
# Changelog
# 


## [0.0.6.2] - 2021-13-11
### Fixed
- Error in panel

## [0.0.6.1] - 2021-01-11
### Fixed
- Environment node uses "Environment" a very often, added this to check for old enviremont textures
- Correct default Saturation values in Node Group

## [0.0.6] - 2021-31-10
### Added
- auto-transfer old environment texture images > prevents from relinking them when switching modes

## [0.0.5] - 2021-30-10
### Changed
- World Control is now created in its own world named "WorldControl"

### Added
- Clean old setup when switching between advanced and basic
- WorldControl has its own world, so we dont mess with user own setup worlds. Easier for cleaning
- Easy switching operator, switch easy between basic and advanced

## [0.0.4] - 2021-30-10
### Changed
- Moved panel to View category
- Panel header > changed dynamic header name. Font size issue, draw_header shows small font?!?

## [0.0.3] - 2021-29-10
### Added
- Reset operator > set all inputs to default
- Panel view easy access from 3d viewport

## [0.0.2] - 2021-28-10
### Fixed 
- The Add function was visible in materials nodes as well, it added empty node in material, added poll to custom nodecategrory 
 
### Changed
- Simplified node names on category, stripped "control"
- NodeGroup width was to narrow, added width

### Added
- Changelog to repo

## [0.0.1] - 2021-27-10
### Added 
- Initial release repo 

# 
#############################################
'''


import os
import bpy
from bpy.types import Operator
from bpy.props import StringProperty
from . import wc_3dv_panel
# from bpy_extras.object_utils import AddObjectHelper


bl_info = {
    "name": "World Control",
    "author": "Rombout Versluijs, Lech Sokolowski (Chocofur)",
    "version": (0, 0, 6,2),
    "blender": (2, 80, 0),
    "location": "World > Add Node > World Control",
    "description": "Adds a shader setup which allows more control when using HDR/EXR lighting. Based on Lech Sokolowski (Chocofur) video BCON19.",
    "doc_url"   : "https://github.com/schroef/World_Control/",
    "tracker_url": "https://github.com/schroef/World_Control/issues",
    "category": "World",
}


class WRD_OT_DeleteWorldControlBasic(Operator):
    """Create a new World add World Control with basic settings"""
    bl_idname = "world.delete_world_control"
    bl_label = "Delete World Control"
    bl_options = {'REGISTER', 'UNDO'}

    def execute(self, context):
        world = bpy.context.scene.world
        nodes = world.node_tree.nodes
        PrincipledWorldWrapper
        principled_world = PrincipledWorldWrapper(world)
        world_output = principled_world.node_out
        for item in nodes:
            if item != world_output:
                # Skip over these when user adds WorldControl from different world 
                # Otherwise we end with an empty fake custom node
                if item.name != "Basic World" and item.name != "Advanced World":
                    nodes.remove(item)

        # self.report({'INFO'}, "Added Basic World Control")
        return {'FINISHED'}


# Source: LilySurfaceSrapper addon
class PrincipledWorldWrapper:
    """This is a wrapper similar in use to PrincipledBSDFWrapper (located in
    bpy_extras.node_shader_utils) but for use with worlds. This is required to
    avoid relying on node names, which depend on Blender's UI language settings
    (see issue #7) """

    def __init__(self, world):
        self.node_background = None
        self.node_out = None
        for n in world.node_tree.nodes:
            if self.node_background is None and n.type == "BACKGROUND":
                self.node_background = n
            elif self.node_out is None and n.type == "OUTPUT_WORLD":
                self.node_out = n

# Source: Thea Render > altered
def get_env_image(context, org_world, img_type):
    '''Search for environment nodes and get image tetures

        :param context: context
        :param org_world: original worl dwhere we look for environemnt texture image
        :param img_type: look what environment image type we need; lgiht or background
        :type image: string
    '''

    # scn = context.scene
    # images = bpy.data.images
    nodes = org_world.node_tree.nodes
    env_node = [item for item in nodes if item.type =='TEX_ENVIRONMENT']

    if env_node:
        for img in env_node:
            # Light HDR
            if img_type == "Light":
                if "Environment Texture" == img.name or "Environment" == img.name:
                    old_img = img.image
            # Background HDR
            if img_type == "Background":
                if "Environment Texture.001" == img.name:
                    old_img = img.image
                    # print(img.name)
                    # print(old_img.name)
                # We are using basic so add this to background when advanced
                else:
                    old_img = img.image

        return old_img

# Source: LilySurfaceScraper
# https://github.com/eliemichel/LilySurfaceScraper/blob/b1e7066f94b965bab90696e0be75faa44eb08c3e/blender/LilySurfaceScraper/CyclesWorldData.py
def getGroundHdriNodeGroup(controlType):
    if ("WorldControl"+controlType) not in bpy.data.node_groups:
        blendfile = os.path.join(os.path.dirname(os.path.realpath(__file__)), "worldcontrol.blend")
        section = "\\NodeTree\\"
        object = "WorldControl" + controlType
        filepath = blendfile + section + object
        directory = blendfile + section
        filename = object

        bpy.ops.wm.append(
            filepath=filepath,
            filename=filename,
            directory=directory)
    return bpy.data.node_groups["WorldControl"+controlType]


def addWorldControl(self, context,controlType):
    scn = bpy.context.scene
    # Create New world
    if bpy.data.worlds.find("WorldControl") == -1:
        world = bpy.data.worlds.new("WorldControl")
    else:
        world = bpy.data.worlds["WorldControl"]    

    # org_world so we can delete fake custom node later
    org_world = bpy.context.scene.world
    scn.world = world
    world.use_nodes = True

    nodes = world.node_tree.nodes
    links = world.node_tree.links

    PrincipledWorldWrapper
    principled_world = PrincipledWorldWrapper(world)
    world_output = principled_world.node_out
    world_loc = world_output.location

    # store old environment images
    old_light_img = get_env_image(context, org_world, img_type="Light")
    old_background_img = get_env_image(context, org_world, img_type="Background")

    # Clear old node setup
    bpy.ops.world.delete_world_control()

    # Add World Control node group
    ground_hdri_node = nodes.new(type="ShaderNodeGroup")
    ground_hdri_node.name = "WorldControl"+controlType
    ground_hdri_node.label = "WorldControl"+controlType
    ground_hdri_node.node_tree = getGroundHdriNodeGroup(controlType)
    ground_hdri_node.width = 185
    ground_hdri_node.location = [world_loc.x - 235, world_loc.y]

    # Shade Node Types
    # https://docs.blender.org/api/current/bpy.types.html
    texture_light_node = nodes.new(type="ShaderNodeTexEnvironment")
    texture_light_node.location = [ground_hdri_node.location.x - 300, world_loc.y]
    texture_light_node.image = old_light_img
    if controlType == "Advanced":
        texture_bg_node = nodes.new(type="ShaderNodeTexEnvironment")
        texture_bg_node.location = [ground_hdri_node.location.x - 300, world_loc.y - 200]
        texture_bg_node.image = old_background_img
    mapping_node = nodes.new(type="ShaderNodeMapping")
    mapping_node.location = [texture_light_node.location.x - 200, world_loc.y]
    texture_coor = nodes.new(type="ShaderNodeTexCoord")
    texture_coor.location = [mapping_node.location.x - 200, world_loc.y]

    # Setup node links
    links.new(texture_coor.outputs["Generated"], mapping_node.inputs["Vector"])
    links.new(mapping_node.outputs[0], texture_light_node.inputs[0])
    links.new(texture_light_node.outputs["Color"], ground_hdri_node.inputs["Light HDR"])
    if controlType == "Advanced":
        links.new(mapping_node.outputs[0], texture_bg_node.inputs[0])
        links.new(texture_bg_node.outputs["Color"], ground_hdri_node.inputs["Background HDR"])
    links.new(ground_hdri_node.outputs[0], world_output.inputs["Surface"])

    # Return orginal world so we clean fake custom node
    scn.world = org_world
    scn.world = world
    return org_world
    

class WRD_OT_AddWorldControlBasic(Operator):
    """Create a new World add World Control with basic settings"""
    bl_idname = "world.add_world_control_basic"
    bl_label = "Add World Control Basic"
    bl_options = {'REGISTER', 'UNDO'}

    def execute(self, context):
        bpy.context.scene.wc_check_mode = "Basic"
        # We use org_world to be able to delete custo node
        org_world = addWorldControl(self,context, "Basic")
        nodes = org_world.node_tree.nodes
        if nodes.find("Basic World") != -1:
            nodes.remove(nodes["Basic World"])
        self.report({'INFO'}, "Added Basic World Control")
        return {'FINISHED'}


class WRD_OT_AddWorldControlAdvanced(Operator):
    """Create a new World add World Control with advanced settings"""
    bl_idname = "world.add_world_control_advanced"
    bl_label = "Add World Control Advanced"
    bl_options = {'REGISTER', 'UNDO'}

    def execute(self, context):
        bpy.context.scene.wc_check_mode = "Advanced"
        # We use org_world to be able to delete custo node
        org_world = addWorldControl(self,context, "Advanced")
        nodes = org_world.node_tree.nodes
        if nodes.find("Advanced World") != -1:
            nodes.remove(nodes["Advanced World"])
        self.report({'INFO'}, "Added Advanced World Control")
        return {'FINISHED'}


class ShaderNodeBasic(bpy.types.NodeCustomGroup):
    bl_name='ShaderNodeBasic'
    bl_label='Basic World'

    def init(self, context):
        # We use org_world to be able to delete custo node
        org_world = addWorldControl(self,context, "Basic")
        nodes = org_world.node_tree.nodes
        nodes.remove(nodes["Basic World"])
        bpy.context.scene.wc_check_mode = "Basic"


class ShaderNodeAdvanced(bpy.types.NodeCustomGroup):
    bl_name='ShaderNodeAdvanced'
    bl_label='Advanced World'

    def init(self, context):
        # We use org_world to be able to delete custo node
        org_world = addWorldControl(self,context, "Advanced")
        nodes = org_world.node_tree.nodes
        nodes.remove(nodes["Advanced World"])
        bpy.context.scene.wc_check_mode = "Advanced"


from nodeitems_utils import NodeItem, register_node_categories, unregister_node_categories
from nodeitems_builtins import ShaderNodeCategory
from nodeitems_utils import NodeCategory, NodeItem


class ShaderWorldNodeCategory(NodeCategory):
    @classmethod
    def poll(cls, context):
        # From Extra Material List addon > MeshLogic
        sdata = context.space_data
        return sdata.shader_type == 'WORLD'


node_categories = [ShaderWorldNodeCategory("WRL_NEW_CUSTOM","World Control",
                                        items=[NodeItem("ShaderNodeBasic",label ='Basic'),
                                                NodeItem("ShaderNodeAdvanced",label ='Advanced')])] 

# Call control mode > allows for easy switching
# Source: https://blender.stackexchange.com/questions/70710/how-to-execute-an-operator-from-an-enumproperty
def update_control_mode(self, context):
    eval('bpy.ops.%s()' % self.wc_switch_control)


classes = [
    ShaderNodeBasic,
    ShaderNodeAdvanced,
    WRD_OT_DeleteWorldControlBasic,
    WRD_OT_AddWorldControlBasic,
    WRD_OT_AddWorldControlAdvanced
]


def register():

    for cls in classes:
        bpy.utils.register_class(cls)
   
    register_node_categories("WCN_CUSTOM_NODES", node_categories)
    wc_3dv_panel.register()
    bpy.types.Scene.wc_check_mode = StringProperty(name="Mode",default="Basic",description="What control type are we using")#,update=wc_3dv_panel.check_world_wc_check_mode)
    bpy.types.Scene.wc_switch_control = bpy.props.EnumProperty(
            name = "Mode",
            description = "Switch contorl mode, switch between Basic and Advanced controls.",
            items = [
                ("world.add_world_control_basic", "Basic", "Basic mode allows for light, diffuse, background and glossy controls"),
                ("world.add_world_control_advanced", "Advanced", "Basic mode allows for light, diffuse, background, reflection and glossy controls. Each items also has both strength and saturation. User can set different background for lighting and background."),              
            ],
            update=update_control_mode
        )

def unregister():
    for cls in classes:
        bpy.utils.unregister_class(cls)
    unregister_node_categories("WCN_CUSTOM_NODES")
    del bpy.types.Scene.wc_check_mode
    wc_3dv_panel.unregister()


if __name__ == "__main__":
    register()
